// charcodes/lib/index.mjs
export var backSpace = 8;
export var tab = 9;
export var lineFeed = 10;
export var carriageReturn = 13;
export var shiftOut = 14;
export var space = 32;
export var exclamationMark = 33;
export var quotationMark = 34;
export var numberSign = 35;
export var dollarSign = 36;
export var percentSign = 37;
export var ampersand = 38;
export var apostrophe = 39;
export var leftParenthesis = 40;
export var rightParenthesis = 41;
export var asterisk = 42;
export var plusSign = 43;
export var comma = 44;
export var dash = 45;
export var dot = 46;
export var slash = 47;
export var digit0 = 48;
export var digit1 = 49;
export var digit2 = 50;
export var digit3 = 51;
export var digit4 = 52;
export var digit5 = 53;
export var digit6 = 54;
export var digit7 = 55;
export var digit8 = 56;
export var digit9 = 57;
export var colon = 58;
export var semicolon = 59;
export var lessThan = 60;
export var equalsTo = 61;
export var greaterThan = 62;
export var questionMark = 63;
export var atSign = 64;
export var uppercaseA = 65;
export var uppercaseB = 66;
export var uppercaseC = 67;
export var uppercaseD = 68;
export var uppercaseE = 69;
export var uppercaseF = 70;
export var uppercaseG = 71;
export var uppercaseH = 72;
export var uppercaseI = 73;
export var uppercaseJ = 74;
export var uppercaseK = 75;
export var uppercaseL = 76;
export var uppercaseM = 77;
export var uppercaseN = 78;
export var uppercaseO = 79;
export var uppercaseP = 80;
export var uppercaseQ = 81;
export var uppercaseR = 82;
export var uppercaseS = 83;
export var uppercaseT = 84;
export var uppercaseU = 85;
export var uppercaseV = 86;
export var uppercaseW = 87;
export var uppercaseX = 88;
export var uppercaseY = 89;
export var uppercaseZ = 90;
export var leftSquareBracket = 91;
export var backslash = 92;
export var rightSquareBracket = 93;
export var caret = 94;
export var underscore = 95;
export var graveAccent = 96;
export var lowercaseA = 97;
export var lowercaseB = 98;
export var lowercaseC = 99;
export var lowercaseD = 100;
export var lowercaseE = 101;
export var lowercaseF = 102;
export var lowercaseG = 103;
export var lowercaseH = 104;
export var lowercaseI = 105;
export var lowercaseJ = 106;
export var lowercaseK = 107;
export var lowercaseL = 108;
export var lowercaseM = 109;
export var lowercaseN = 110;
export var lowercaseO = 111;
export var lowercaseP = 112;
export var lowercaseQ = 113;
export var lowercaseR = 114;
export var lowercaseS = 115;
export var lowercaseT = 116;
export var lowercaseU = 117;
export var lowercaseV = 118;
export var lowercaseW = 119;
export var lowercaseX = 120;
export var lowercaseY = 121;
export var lowercaseZ = 122;
export var leftCurlyBrace = 123;
export var verticalBar = 124;
export var rightCurlyBrace = 125;
export var tilde = 126;
export var nonBreakingSpace = 160;
export var oghamSpaceMark = 5760;
export var lineSeparator = 8232;
export var paragraphSeparator = 8233;
export function isDigit(code) {
  return code >= digit0 && code <= digit9;
}
